chain/network/src/peer/peer_actor.rs

```diff
@@ -247,7 +236,61 @@ impl PeerActor {
                 network_state,
                 connection: None,
                 _connection_guard: connection_guard,
-            }
+            };
+            let addr = ctx.address();
+            let stats = pa.stats.clone();
+            // Event loop receiving and processing messages.
+            // After dispatching a message, loop starts parsing the next message immediately
+            // (i.e. it doesn't wait for the message to be processed).
+            // It uses a fixed small buffer allocated by BufReader.
+            // For each message it allocates a Vec with exact size of the message.
+            // TODO(gprusak): once borsh support is dropped, we can parse a proto
+            // directly from the stream.
+            ctx.spawn(
+                async move {
+                    const READ_BUFFER_CAPACITY: usize = 8 * 1024;
+                    let mut read = tokio::io::BufReader::with_capacity(READ_BUFFER_CAPACITY, read);
+                    loop {
+                        let n =
+                            read.read_u32_le().await.map_err(RecvLoopError::Receiving)? as usize;
+                        if n > NETWORK_MESSAGE_MAX_SIZE_BYTES {
+                            addr.do_send(RecvMessage(Err(ReasonForBan::Abusive)));
+                            return Err(RecvLoopError::MessageTooLarge {
+                                got_bytes: n,
+                                want_max_bytes: NETWORK_MESSAGE_MAX_SIZE_BYTES,
+                            });
+                        }
+                        let mut buf = Vec::new();
+                        buf.resize(n, 0);
+                        read.read_exact(&mut buf[..]).await.map_err(RecvLoopError::Receiving)?;
```

Comment:
done

Reply:
Sorry, I was unclear -- I was thinking about "size of the buffer" metric rather than "processing time". Basically, if ten peers send us 100 megs each, it would be nice to observe that in the metric.


Specifically, either we should set `PEER_DATA_READ_BUFFER_SIZE` here, or we should delete that metric altogether (as it is unused as of this PR).

Reply:
I've replaced PEER_DATA_READ_BUFFER_SIZE with a small histogram of message sizes, because it wouldn't make much sense to have a gauge tracking the size of the last message, I think.